Mathlib/Topology/Instances/CantorSet.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2024 Jana Göken. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Artur Szafarczyk, Suraj Krishna M S, JB Stiegler, Isabelle Dubois,
+Tomáš Jakl, Lorenzo Zanichelli, Alina Yan, Emilie Uthaiwat, Jana Göken
+under guidance of Filippo A. E. Nuccio
+-/
+import Mathlib.Topology.Metrizable.Basic
+import Mathlib.Topology.Algebra.GroupWithZero
+import Mathlib.Topology.Instances.Real
+
+/-!
+# Ternary Cantor Set
+This file contains the definition of the Cantor ternary set
+as well as some first properties, which will be updated later.
+
+## Main Definitions
+We give a definition for the ternary Cantor set in this file, which
+is `cantorSet`.
+-/
+
+/-- We define the preCantorSet as the preimages under the iterations of some functions. -/
+def preCantorSet : ℕ → Set ℝ
+  | 0 => Set.Icc 0 1
+  | Nat.succ n => (· / 3) '' preCantorSet n ∪ (fun x ↦ (2 + x) / 3)  '' preCantorSet n
+
```

Comment:
As Riccardo indicated, the following two theorems will make your life easier:
```suggestion

@[simp] lemma preCantorSet_zero : preCantorSet 0 = Set.Icc 0 1 := rfl
@[simp] lemma preCantorSet_succ (n : ℕ) :
    preCantorSet (n + 1) = (· / 3) '' preCantorSet n ∪ (fun x ↦ (2 + x) / 3) '' preCantorSet n :=
  rfl
```

Reply:
Thank you, I finally understand. Very interesting that this actually helps with anything.

Reply:
If you want something to look up, those are called "equational lemmas". Lean in fact knows how to generate them, and this is what `simp` rewrites with when you write `simp [preCantorSet]` (in particular, it does *not* just unfold the definition of `preCantorSet`), however it is good practice to write them yourself as
1. Lean is kind of inconsistent as to how those equational lemmas should look like, eg in a recent Lean 4 version there would be a single equational lemma of the form `preCantorSet n = match n with | 0 => Set.Icc 0 1 | n + 1 => ...` instead of the pair `preCantorSet_zero`, `preCantorSet_succ` that I wrote about
2. If your definition ever changes, the autogenerated equational lemmas will too, which is undesirable from an API stability perspective

Reply:
understood, thanks!